File: intellij/src/test/kotlin/ai/jolli/jollimemory/core/SummaryTreeTest.kt
package ai.jolli.jollimemory.core

import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.kotest.matchers.string.shouldContain
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class SummaryTreeTest {

    private fun makeTopic(title: String = "Test Topic") = TopicSummary(
        title = title,
        trigger = "Test trigger",
        response = "Test response",
        decisions = "Test decisions",
    )

    private fun makeLeaf(
        hash: String = "abc123",
        topics: List<TopicSummary> = listOf(makeTopic()),
        stats: DiffStats = DiffStats(3, 100, 50),
        turns: Int = 5,
        date: String = "2026-01-15T10:00:00Z",
    ) = CommitSummary(
        commitHash = hash,
        commitMessage = "Test commit",
        commitAuthor = "Alice",
        commitDate = date,
        branch = "main",
        generatedAt = "2026-01-15T10:00:00Z",
        topics = topics,
        stats = stats,
        conversationTurns = turns,
    )

    private fun makeSquash(children: List<CommitSummary>) = CommitSummary(
        commitHash = "squash123",
        commitMessage = "Squash commit",
        commitAuthor = "Bob",
        commitDate = "2026-01-20T10:00:00Z",
        branch = "main",
        generatedAt = "2026-01-20T10:00:00Z",
        commitType = CommitType.squash,
        children = children,
    )

    // ── collectAllTopics ────────────────────────────────────────────────

    @Nested
    inner class CollectAllTopics {
        @Test
        fun `collects topics from leaf node`() {
            val node = makeLeaf(topics = listOf(makeTopic("A"), makeTopic("B")))
            val result = SummaryTree.collectAllTopics(node)
            result shouldHaveSize 2
            result[0].topic.title shouldBe "A"
            result[1].topic.title shouldBe "B"
        }

        @Test
        fun `collects topics from children first, then own`() {
            val child1 = makeLeaf(hash = "c1", topics = listOf(makeTopic("Child1")), date = "2026-01-10T10:00:00Z")
            val child2 = makeLeaf(hash = "c2", topics = listOf(makeTopic("Child2")), date = "2026-01-12T10:00:00Z")
            val parent = makeSquash(listOf(child1, child2))

            val result = SummaryTree.collectAllTopics(parent)
            // Children are reversed (oldest first), so child2 first (it's second in list, reversed = first)
            result shouldHaveSize 2
            result[0].topic.title shouldBe "Child2"
            result[1].topic.title shouldBe "Child1"
        }

        @Test
        fun `returns empty for node without topics`() {
            val node = makeLeaf(topics = emptyList())
            SummaryTree.collectAllTopics(node).shouldHaveSize(0)
        }
    }

    // ── aggregateStats ──────────────────────────────────────────────────

    @Nested
    inner class AggregateStats {
        @Test
        fun `returns own stats for leaf node`() {
            val node = makeLeaf(stats = DiffStats(5, 200, 100))
            val result = SummaryTree.aggregateStats(node)
            result.filesChanged shouldBe 5
            result.insertions shouldBe 200
            result.deletions shouldBe 100
        }

        @Test
        fun `sums stats across children`() {
            val child1 = makeLeaf(hash = "c1", stats = DiffStats(2, 50, 10))
            val child2 = makeLeaf(hash = "c2", stats = DiffStats(3, 75, 25))
            val parent = makeSquash(listOf(child1, child2))

            val result = SummaryTree.aggregateStats(parent)
            result.filesChanged shouldBe 5
            result.insertions shouldBe 125
            result.deletions shouldBe 35
        }

        @Test
        fun `handles null stats`() {
            val node = CommitSummary(
                commitHash = "x", commitMessage = "x", commitAuthor = "x",
                commitDate = "2026-01-01T00:00:00Z", branch = "main",
                generatedAt = "2026-01-01T00:00:00Z",
            )
            val result = SummaryTree.aggregateStats(node)
            result.filesChanged shouldBe 0
            result.insertions shouldBe 0
            result.deletions shouldBe 0
        }
    }

    // ── aggregateTurns ──────────────────────────────────────────────────

    @Nested
    inner class AggregateTurns {
        @Test
        fun `returns own turns for leaf`() {
            SummaryTree.aggregateTurns(makeLeaf(turns = 7)) shouldBe 7
        }

        @Test
        fun `sums turns across tree`() {
            val child1 = makeLeaf(hash = "c1", turns = 3)
            val child2 = makeLeaf(hash = "c2", turns = 4)
            val parent = makeSquash(listOf(child1, child2))
            SummaryTree.aggregateTurns(parent) shouldBe 7
        }
    }

    // ── aggregateConversationTokens / Breakdown / EstimatedCost ─────────

    @Nested
    inner class AggregateConversationUsage {
        @Test
        fun `sums tokens, breakdown, and cost across the tree`() {
            val c1 = makeLeaf(hash = "c1").copy(
                conversationTokens = 100,
                conversationTokenBreakdown = ConversationTokenBreakdown(60, 30, 10),
                estimatedCostUsd = 0.25,
            )
            val c2 = makeLeaf(hash = "c2").copy(
                conversationTokens = 200,
                conversationTokenBreakdown = ConversationTokenBreakdown(120, 60, 20),
                estimatedCostUsd = 0.50,
            )
            val root = makeSquash(listOf(c1, c2))
            SummaryTree.aggregateConversationTokens(root) shouldBe 300L
            SummaryTree.aggregateConversationTokenBreakdown(root) shouldBe ConversationTokenBreakdown(180, 90, 30)
            SummaryTree.aggregateEstimatedCost(root) shouldBe 0.75
        }

        @Test
        fun `returns zeros when no node carries usage`() {
            val root = makeSquash(listOf(makeLeaf(hash = "c1")))
            SummaryTree.aggregateConversationTokens(root) shouldBe 0L
            SummaryTree.aggregateEstimatedCost(root) shouldBe 0.0
            SummaryTree.aggregateConversationTokenBreakdown(root) shouldBe ConversationTokenBreakdown(0, 0, 0)
        }

        @Test
        fun `aggregateEstimatedCost is a pure stored sum with no fallback`() {
            // The Sonnet/model fallback lives in the display layer (matching VS Code), so the
            // aggregate itself stays pure: a node with models/breakdown but no stored cost is 0.
            val node = makeLeaf().copy(
                conversationTokens = 3_000_000,
                conversationTokenBreakdown = ConversationTokenBreakdown(1_000_000, 1_000_000, 1_000_000),
                conversationModels = listOf(
                    ModelTokenUsage("claude-opus-4-8", ModelPricing.providerOf("claude-opus-4-8"), output = 1_000_000),
                ),
                estimatedCostUsd = null,
            )
            SummaryTree.aggregateEstimatedCost(node) shouldBe 0.0
        }

        @Test
        fun `falls back to legacy tokenUsage for breakdown and total`() {
            val node = makeLeaf().copy(
                conversationTokens = null,
                conversationTokenBreakdown = null,
                tokenUsage = TokenUsage(inputTokens = 100, outputTokens = 50, cacheWriteTokens = 10),
            )
            SummaryTree.aggregateConversationTokenBreakdown(node) shouldBe ConversationTokenBreakdown(100, 50, 10)
            SummaryTree.aggregateConversationTokens(node) shouldBe 160L
        }
    }

    // ── countTopics ─────────────────────────────────────────────────────

    @Nested
    inner class CountTopics {
        @Test
        fun `counts own topics`() {
            SummaryTree.countTopics(makeLeaf(topics = listOf(makeTopic(), makeTopic()))) shouldBe 2
        }

        @Test
        fun `counts across tree`() {
            val child1 = makeLeaf(hash = "c1", topics = listOf(makeTopic()))
            val child2 = makeLeaf(hash = "c2", topics = listOf(makeTopic(), makeTopic()))
            val parent = makeSquash(listOf(child1, child2))
            SummaryTree.countTopics(parent) shouldBe 3
        }
    }

    // ── collectSourceNodes ──────────────────────────────────────────────

    @Nested
    inner class CollectSourceNodes {
        @Test
        fun `returns leaf with topics`() {
            val node = makeLeaf()
            val result = SummaryTree.collectSourceNodes(node)
            result shouldHaveSize 1
        }

        @Test
        fun `skips nodes without topics`() {
            val node = makeLeaf(topics = emptyList())
            SummaryTree.collectSourceNodes(node).shouldHaveSize(0)
        }

        @Test
        fun `collects from children`() {
            val child1 = makeLeaf(hash = "c1")
            val child2 = makeLeaf(hash = "c2")
            val parent = makeSquash(listOf(child1, child2))
            SummaryTree.collectSourceNodes(parent) shouldHaveSize 2
        }
    }

    // ── isLeafNode ──────────────────────────────────────────────────────

    @Test
    fun `isLeafNode returns true for node without children`() {
        SummaryTree.isLeafNode(makeLeaf()) shouldBe true
    }

    @Test
    fun `isLeafNode returns false for node with children`() {
        val parent = makeSquash(listOf(makeLeaf()))
        SummaryTree.isLeafNode(parent) shouldBe false
    }

    // ── computeDurationDays ─────────────────────────────────────────────

    @Test
    fun `computeDurationDays returns 1 for single source`() {
        SummaryTree.computeDurationDays(makeLeaf()) shouldBe 1
    }

    @Test
    fun `computeDurationDays counts distinct dates`() {
        val child1 = makeLeaf(hash = "c1", date = "2026-01-10T10:00:00Z")
        val child2 = makeLeaf(hash = "c2", date = "2026-01-12T10:00:00Z")
        val child3 = makeLeaf(hash = "c3", date = "2026-01-12T15:00:00Z") // same day as child2
        val parent = makeSquash(listOf(child1, child2, child3))
        SummaryTree.computeDurationDays(parent) shouldBe 2
    }

    // ── formatDurationLabel ─────────────────────────────────────────────

    @Test
    fun `formatDurationLabel shows day count and date range`() {
        val child1 = makeLeaf(hash = "c1", date = "2026-01-10T10:00:00Z")
        val child2 = makeLeaf(hash = "c2", date = "2026-01-12T10:00:00Z")
        val parent = makeSquash(listOf(child1, child2))
        val label = SummaryTree.formatDurationLabel(parent)
        label shouldContain "2 days"
        label shouldContain "Jan"
    }

    @Test
    fun `formatDurationLabel shows 1 day for single source`() {
        SummaryTree.formatDurationLabel(makeLeaf()) shouldBe "1 day"
    }

    // ── updateTopicInTree ───────────────────────────────────────────────

    @Nested
    inner class UpdateTopicInTree {
        @Test
        fun `updates topic at index 0 in leaf`() {
            val node = makeLeaf(topics = listOf(makeTopic("Original")))
            val updates = TopicUpdates(title = "Updated")
            val result = SummaryTree.updateTopicInTree(node, 0, updates)

            result shouldNotBe null
            result!!.result.topics!![0].title shouldBe "Updated"
        }

        @Test
        fun `preserves unchanged fields`() {
            val original = TopicSummary("Title", "Trigger", "Response", "Decisions", todo = "Todo")
            val node = makeLeaf(topics = listOf(original))
            val updates = TopicUpdates(title = "New Title")
            val result = SummaryTree.updateTopicInTree(node, 0, updates)!!

            val updated = result.result.topics!![0]
            updated.title shouldBe "New Title"
            updated.trigger shouldBe "Trigger"
            updated.response shouldBe "Response"
            updated.decisions shouldBe "Decisions"
            updated.todo shouldBe "Todo"
        }

        @Test
        fun `updates topic in child node`() {
            val child = makeLeaf(hash = "c1", topics = listOf(makeTopic("Child Topic")))
            val parent = makeSquash(listOf(child))
            val updates = TopicUpdates(title = "Updated Child")
            val result = SummaryTree.updateTopicInTree(parent, 0, updates)!!

            result.result.children!![0].topics!![0].title shouldBe "Updated Child"
        }
    }

    // ── deleteTopicInTree ───────────────────────────────────────────────

    @Nested
    inner class DeleteTopicInTree {
        @Test
        fun `deletes topic at index 0`() {
            val node = makeLeaf(topics = listOf(makeTopic("A"), makeTopic("B")))
            val result = SummaryTree.deleteTopicInTree(node, 0)!!
            result.result.topics!! shouldHaveSize 1
            result.result.topics!![0].title shouldBe "B"
        }

        @Test
        fun `deletes topic from child node`() {
            val child = makeLeaf(hash = "c1", topics = listOf(makeTopic("Only")))
            val parent = makeSquash(listOf(child))
            val result = SummaryTree.deleteTopicInTree(parent, 0)!!
            result.result.children!![0].topics!! shouldHaveSize 0
        }
    }
}

File: cli/src/core/SummaryTree.test.ts
import { describe, expect, it } from "vitest";
import type { CommitSummary } from "../Types.js";
import {
	aggregateStats,
	aggregateTurns,
	collectAllTopics,
	collectAllTranscriptHashes,
	collectDisplayTopics,
	collectSourceNodes,
	computeDurationDays,
	countTopics,
	deleteTopicInTree,
	formatDurationLabel,
	isLeafNode,
	isUnifiedHoistFormat,
	resolveDiffStats,
	updateTopicInTree,
} from "./SummaryTree.js";

/** Helper to create a minimal leaf node */
function leaf(overrides: Partial<CommitSummary> = {}): CommitSummary {
	return {
		version: 3,
		commitHash: "aaa",
		commitMessage: "msg",
		commitAuthor: "author",
		commitDate: "2026-03-01T10:00:00.000Z",
		branch: "main",
		generatedAt: "2026-03-01T10:00:10.000Z",
		transcriptEntries: 5,
		conversationTurns: 3,
		llm: { model: "test-model", inputTokens: 100, outputTokens: 50, apiLatencyMs: 1000, stopReason: "end_turn" },
		stats: { filesChanged: 2, insertions: 50, deletions: 10 },
		topics: [{ title: "Topic A", trigger: "t", response: "r", decisions: "d" }],
		...overrides,
	};
}

// ─── Scenario setup ──────────────────────────────────────────────────────────
// A = original commit (10:00)
// C = amend of A (10:05) — has own topics + children:[A]
// B = independent commit (10:20)
// D = squash of C+B (10:30) — pure container, children:[B, C] (newest first)

const A = leaf({
	commitHash: "aaa",
	commitMessage: "feat: login",
	commitDate: "2026-03-01T10:00:00.000Z",
	stats: { filesChanged: 3, insertions: 120, deletions: 5 },
	conversationTurns: 8,
	topics: [{ title: "Implement login", trigger: "t1", response: "r1", decisions: "d1" }],
});

const C: CommitSummary = {
	version: 3,
	commitHash: "ccc",
	commitMessage: "feat: login",
	commitAuthor: "author",
	commitDate: "2026-03-01T10:05:00.000Z",
	branch: "main",
	generatedAt: "2026-03-01T10:05:10.000Z",
	transcriptEntries: 2,
	conversationTurns: 1,
	stats: { filesChanged: 1, insertions: 3, deletions: 3 },
	topics: [{ title: "Fix indentation", trigger: "t2", response: "r2", decisions: "d2" }],
	children: [A],
};

const B = leaf({
	commitHash: "bbb",
	commitMessage: "feat: logout",
	commitDate: "2026-03-01T10:20:00.000Z",
	stats: { filesChanged: 2, insertions: 80, deletions: 2 },
	conversationTurns: 5,
	topics: [{ title: "Implement logout", trigger: "t3", response: "r3", decisions: "d3" }],
});

const D: CommitSummary = {
	version: 3,
	commitHash: "ddd",
	commitMessage: "feat: auth module",
	commitAuthor: "author",
	commitDate: "2026-03-01T10:30:00.000Z",
	branch: "main",
	generatedAt: "2026-03-01T10:30:03.000Z",
	// Pure container: no topics, stats, llm
	children: [B, C], // newest first
};

describe("SummaryTree", () => {
	describe("collectAllTopics", () => {
		it("returns own topics for a leaf node", () => {
			const topics = collectAllTopics(A);
			expect(topics).toHaveLength(1);
			expect(topics[0].title).toBe("Implement login");
			expect(topics[0].commitDate).toBe("2026-03-01T10:00:00.000Z");
		});

		it("returns children topics before own for an amend node", () => {
			const topics = collectAllTopics(C);
			expect(topics).toHaveLength(2);
			expect(topics[0].title).toBe("Implement login"); // A (older)
			expect(topics[1].title).toBe("Fix indentation"); // C own (newer)
		});

		it("returns all topics in chronological order for a squash node", () => {
			const topics = collectAllTopics(D);
			expect(topics).toHaveLength(3);
			// D.children = [B, C], reversed → [C, B] for chronological
			// C recurse: A(10:00) then C(10:05), then B(10:20)
			expect(topics[0].title).toBe("Implement login"); // A
			expect(topics[1].title).toBe("Fix indentation"); // C
			expect(topics[2].title).toBe("Implement logout"); // B
		});

		it("returns empty array for a container with no topics", () => {
			const empty: CommitSummary = {
				...D,
				children: [],
			};
			expect(collectAllTopics(empty)).toHaveLength(0);
		});
	});

	describe("aggregateStats", () => {
		it("returns own stats for a leaf node", () => {
			const stats = aggregateStats(A);
			expect(stats).toEqual({ filesChanged: 3, insertions: 120, deletions: 5 });
		});

		it("aggregates across amend tree", () => {
			const stats = aggregateStats(C);
			expect(stats).toEqual({ filesChanged: 4, insertions: 123, deletions: 8 });
		});

		it("aggregates across full squash tree", () => {
			const stats = aggregateStats(D);
			// D(0) + B(2,80,2) + C(1,3,3) + A(3,120,5)
			expect(stats).toEqual({ filesChanged: 6, insertions: 203, deletions: 10 });
		});

		it("does not mutate the original node stats", () => {
			const before = JSON.stringify(A.stats);
			aggregateStats(D);
			expect(JSON.stringify(A.stats)).toBe(before);
		});

		it("handles missing stats", () => {
			const noStats: CommitSummary = { ...D, stats: undefined };
			const stats = aggregateStats(noStats);
			// Only children contribute
			expect(stats.insertions).toBe(203);
		});
	});

	describe("aggregateTurns", () => {
		it("returns own turns for a leaf", () => {
			expect(aggregateTurns(A)).toBe(8);
		});

		it("sums across tree", () => {
			// D(0) + B(5) + C(1) + A(8) = 14
			expect(aggregateTurns(D)).toBe(14);
		});
	});

	describe("countTopics", () => {
		it("counts leaf topics", () => {
			expect(countTopics(A)).toBe(1);
		});

		it("counts across full tree", () => {
			expect(countTopics(D)).toBe(3);
		});
	});

	describe("collectSourceNodes (leaf-only, root excluded)", () => {
		it("returns empty for a leaf node (root itself is not a source of itself)", () => {
			expect(collectSourceNodes(A)).toEqual([]);
		});

		it("returns the leaf child of an amend root (root excluded)", () => {
			const nodes = collectSourceNodes(C);
			expect(nodes.map((n) => n.commitHash)).toEqual(["aaa"]);
		});

		it("returns leaf descendants of a squash root, skipping intermediate amend container", () => {
			// D.children = [B, C]; B is a leaf so it's included; C is an intermediate
			// container (has children), so we recurse into C and pick up A as a leaf.
			const nodes = collectSourceNodes(D);
			expect(nodes.map((n) => n.commitHash)).toEqual(["bbb", "aaa"]);
		});
	});

	describe("isUnifiedHoistFormat", () => {
		it("returns false for v3 (legacy) summaries", () => {
			expect(isUnifiedHoistFormat({ version: 3 })).toBe(false);
		});

		it("returns true for v4 (unified Hoist) summaries", () => {
			expect(isUnifiedHoistFormat({ version: 4 })).toBe(true);
		});

		it("treats future versions as unified Hoist (forward-compatible)", () => {
			expect(isUnifiedHoistFormat({ version: 5 })).toBe(true);
			expect(isUnifiedHoistFormat({ version: 99 })).toBe(true);
		});

		it("returns false when version is below 4 (defensive)", () => {
			expect(isUnifiedHoistFormat({ version: 0 })).toBe(false);
			expect(isUnifiedHoistFormat({ version: 1 })).toBe(false);
		});
	});

	describe("collectDisplayTopics", () => {
		const v4Root = {
			...A,
			version: 4,
			topics: [
				{ title: "Topic X", trigger: "tx", response: "rx", decisions: "dx" },
				{ title: "Topic Y", trigger: "ty", response: "ry", decisions: "dy" },
			],
			children: [
				// Stripped child (no topics) -- v4 strip removes them
				{ ...A, version: 3, topics: [] as ReadonlyArray<never>, commitHash: "child-aaa" },
			],
		};

		it("returns root.topics for v4 (root authoritative, ignore children)", () => {
			const topics = collectDisplayTopics(v4Root);
			expect(topics.map((t) => t.title)).toEqual(["Topic X", "Topic Y"]);
		});

		it("returns empty array for v4 recap-only commit (topics=[] is legitimate)", () => {
			const recapOnly = { ...v4Root, topics: [], recap: "Just a recap" };
			expect(collectDisplayTopics(recapOnly)).toEqual([]);
		});

		it("falls back to recursive collectAllTopics for v3 (legacy)", () => {
			// D is v3 with no own topics, children carry the topics
			const topics = collectDisplayTopics(D);
			expect(topics).toHaveLength(3);
			expect(topics.map((t) => t.title)).toEqual(["Implement login", "Fix indentation", "Implement logout"]);
		});

		it("decorates v4 topics with commitDate / generatedAt from root", () => {
			const topics = collectDisplayTopics(v4Root);
			expect(topics[0].commitDate).toBe(v4Root.commitDate);
			expect(topics[0].generatedAt).toBe(v4Root.generatedAt);
		});
	});

	describe("collectAllTranscriptHashes", () => {
		it("returns just the root hash for a leaf", () => {
			expect(collectAllTranscriptHashes(A)).toEqual(["aaa"]);
		});

		it("walks children depth-first, root first", () => {
			expect(collectAllTranscriptHashes(C)).toEqual(["ccc", "aaa"]);
		});

		it("descends into nested squash trees", () => {
			expect(collectAllTranscriptHashes(D)).toEqual(["ddd", "bbb", "ccc", "aaa"]);
		});
	});

	describe("isLeafNode", () => {
		it("returns true for node without children", () => {
			expect(isLeafNode(A)).toBe(true);
		});

		it("returns false for node with children", () => {
			expect(isLeafNode(C)).toBe(false);
		});

		it("returns true for empty children array", () => {
			expect(isLeafNode({ ...A, children: [] })).toBe(true);
		});
	});

	describe("computeDurationDays", () => {
		it("returns 1 for a leaf node", () => {
			expect(computeDurationDays(A)).toBe(1);
		});

		it("returns 1 when all sources are same day", () => {
			expect(computeDurationDays(D)).toBe(1); // all on 2026-03-01
		});

		it("returns correct days for multi-day span", () => {
			const multiDay: CommitSummary = {
				...D,
				children: [
					{ ...B, commitDate: "2026-03-03T10:00:00.000Z", generatedAt: "2026-03-03T10:00:10.000Z" },
					{ ...A, commitDate: "2026-03-01T10:00:00.000Z", generatedAt: "2026-03-01T10:00:10.000Z" },
				],
			};
			expect(computeDurationDays(multiDay)).toBe(2);
		});

		it("falls back to commitDate when a source has no generatedAt (covers `||` right branch)", () => {
			// One child with empty generatedAt forces the fallback path inside the Set map.
			const mixed: CommitSummary = {
				...D,
				children: [
					{ ...B, commitDate: "2026-03-03T10:00:00.000Z", generatedAt: "" },
					{ ...A, commitDate: "2026-03-01T10:00:00.000Z", generatedAt: "2026-03-01T10:00:10.000Z" },
				],
			};
			expect(computeDurationDays(mixed)).toBe(2);
		});
	});

	describe("formatDurationLabel", () => {
		it("returns '1 day' for leaf node", () => {
			expect(formatDurationLabel(A)).toBe("1 day");
		});

		it("includes date range for multi-source same day", () => {
			const label = formatDurationLabel(D);
			// Multiple sources even on same day → shows range
			expect(label).toMatch(/^1 day \(/);
		});

		it("includes date range for multi-day span", () => {
			const multiDay: CommitSummary = {
				...D,
				children: [
					{ ...B, commitDate: "2026-03-05T10:00:00.000Z", generatedAt: "2026-03-05T10:00:10.000Z" },
					{ ...A, commitDate: "2026-03-01T10:00:00.000Z", generatedAt: "2026-03-01T10:00:10.000Z" },
				],
			};
			const label = formatDurationLabel(multiDay);
			expect(label).toMatch(/^2 days \(/);
			expect(label).toContain("Mar");
		});

		it("falls back to commitDate for timestamp computation when generatedAt is empty", () => {
			// Covers the `s.generatedAt || s.commitDate` right branch in formatDurationLabel.
			const mixed: CommitSummary = {
				...D,
				children: [
					{ ...B, commitDate: "2026-03-05T10:00:00.000Z", generatedAt: "" },
					{ ...A, commitDate: "2026-03-01T10:00:00.000Z", generatedAt: "2026-03-01T10:00:10.000Z" },
				],
			};
			const label = formatDurationLabel(mixed);
			expect(label).toMatch(/^2 days \(/);
		});
	});

	describe("updateTopicInTree", () => {
		it("updates a leaf topic by global index", () => {
			const result = updateTopicInTree(A, 0, { title: "Updated login" });
			expect(result).not.toBeNull();
			expect(result?.consumed).toBe(1);
			expect(result?.result.topics?.[0]?.title).toBe("Updated login");
			expect(A.topics?.[0]?.title).toBe("Implement login");
		});

		it("updates the oldest child topic before newer siblings", () => {
			const result = updateTopicInTree(D, 0, { decisions: "changed" });
			expect(result).not.toBeNull();
			const updatedTopics = collectAllTopics((result as NonNullable<typeof result>).result);
			expect(updatedTopics[0].title).toBe("Implement login");
			expect(updatedTopics[0].decisions).toBe("changed");
			expect(updatedTopics[1].decisions).toBe("d2");
			expect(updatedTopics[2].decisions).toBe("d3");
		});

		it("updates own topics after child topics have been consumed", () => {
			const result = updateTopicInTree(C, 1, { response: "new-response" });
			expect(result).not.toBeNull();
			expect(result?.consumed).toBe(2);
			expect(result?.result.topics?.[0]?.response).toBe("new-response");
			expect(result?.result.children?.[0]?.topics?.[0]?.response).toBe("r1");
		});

		it("updates only the targeted topic when a node has multiple own topics", () => {
			const multiTopicNode: CommitSummary = {
				...A,
				topics: [
					{ title: "Topic A", trigger: "t1", response: "r1", decisions: "d1" },
					{ title: "Topic B", trigger: "t2", response: "r2", decisions: "d2" },
				],
			};

			const result = updateTopicInTree(multiTopicNode, 1, { decisions: "updated" });

			expect(result).not.toBeNull();
			expect(result?.result.topics?.[0]?.decisions).toBe("d1");
			expect(result?.result.topics?.[1]?.decisions).toBe("updated");
		});

		it("returns the original tree unchanged when the global index is out of range", () => {
			const result = updateTopicInTree(D, 99, { title: "nope" });
			expect(result).not.toBeNull();
			expect(result?.consumed).toBe(3);
			expect(result?.result).toBe(D);
		});
	});

	describe("deleteTopicInTree", () => {
		it("deletes a leaf topic by global index", () => {
			const result = deleteTopicInTree(A, 0);
			expect(result).not.toBeNull();
			expect(result?.consumed).toBe(1);
			expect(result?.result.topics).toEqual([]);
			expect(A.topics).toHaveLength(1);
		});

		it("deletes an own topic after traversing child topics first", () => {
			const result = deleteTopicInTree(C, 1);
			expect(result).not.toBeNull();
			expect(result?.result.topics).toEqual([]);
			expect(result?.result.children?.[0]?.topics?.[0]?.title).toBe("Implement login");
		});

		it("deletes a child topic while preserving newer siblings", () => {
			const result = deleteTopicInTree(D, 0);
			expect(result).not.toBeNull();
			const remainingTopics = collectAllTopics((result as NonNullable<typeof result>).result);
			expect(remainingTopics.map((topic) => topic.title)).toEqual(["Fix indentation", "Implement logout"]);
			expect(result?.result.children?.[0]?.commitHash).toBe("bbb");
		});

		it("returns the original tree unchanged when deleting an out-of-range topic", () => {
			const result = deleteTopicInTree(D, 99);
			expect(result).not.toBeNull();
			expect(result?.consumed).toBe(3);
			expect(result?.result).toBe(D);
		});
	});

	describe("resolveDiffStats", () => {
		it("returns diffStats when present (new-data path) — ignores stats and children", () => {
			const node: CommitSummary = {
				...A,
				diffStats: { filesChanged: 99, insertions: 999, deletions: 9 },
				stats: { filesChanged: 1, insertions: 1, deletions: 1 },
				children: [B],
			};
			expect(resolveDiffStats(node)).toEqual({ filesChanged: 99, insertions: 999, deletions: 9 });
		});

		it("returns node.stats for a leaf (no children, no diffStats)", () => {
			// A is a leaf with stats but no diffStats
			expect(resolveDiffStats(A)).toEqual({ filesChanged: 3, insertions: 120, deletions: 5 });
		});

		it("returns zeros for a leaf with neither diffStats nor stats", () => {
			const empty: CommitSummary = { ...A, stats: undefined };
			expect(resolveDiffStats(empty)).toEqual({ filesChanged: 0, insertions: 0, deletions: 0 });
		});

		it("treats undefined children and [] children both as leaf", () => {
			const withEmptyChildren: CommitSummary = { ...A, children: [] };
			expect(resolveDiffStats(withEmptyChildren)).toEqual({ filesChanged: 3, insertions: 120, deletions: 5 });
		});

		it("aggregates across tree for a container (legacy squash root: no diffStats, no stats, has children)", () => {
			// D is a pure container with children [B, C] where C in turn has child A.
			// Matches today's aggregateStats behavior exactly.
			const viaHelper = resolveDiffStats(D);
			const viaAggregate = aggregateStats(D);
			expect(viaHelper).toEqual(viaAggregate);
			// Sanity: the sum matches aggregateStats's own test expectation
			expect(viaHelper).toEqual({ filesChanged: 6, insertions: 203, deletions: 10 });
		});

		it("aggregates for a legacy amend root (stats is delta, has children)", () => {
			// Legacy amend: stats = delta (small), children carry pre-amend full stats.
			// Today's display is aggregateStats(amend) = delta + children.stats. Must
			// stay identical under resolveDiffStats for backward compatibility.
			const legacyAmend: CommitSummary = {
				...C, // C already has stats (delta-ish) + children [A]
			};
			expect(resolveDiffStats(legacyAmend)).toEqual(aggregateStats(legacyAmend));
			expect(resolveDiffStats(legacyAmend)).toEqual({ filesChanged: 4, insertions: 123, deletions: 8 });
		});

		it("aggregates for a legacy nested container (no diffStats anywhere, no root stats, has children)", () => {
			// Pure nested container: mimics an old squash/rebase-pick root viewed via getSummary.
			const nested: CommitSummary = {
				...D,
				stats: undefined,
				diffStats: undefined,
				children: [B, C], // reuse same sub-tree
			};
			expect(resolveDiffStats(nested)).toEqual(aggregateStats(nested));
		});

		it("returns diffStats even when children are present (new container root)", () => {
			const newSquashRoot: CommitSummary = {
				...D,
				diffStats: { filesChanged: 2, insertions: 40, deletions: 20 },
				// No stats (container), children still present
				children: [B, C],
			};
			expect(resolveDiffStats(newSquashRoot)).toEqual({
				filesChanged: 2,
				insertions: 40,
				deletions: 20,
			});
		});
	});
});
